platform/core/src/services/CustomizationService/CustomizationService.ts

```diff
@@ -235,292 +173,333 @@ export default class CustomizationService extends PubSubService {
-      const extensionValue = this.findExtensionValue(value);
-      // The child of a reference is only a set of references when an array,
-      // so call the addReference direct.  It could be a secondary reference perhaps
-      this.addReference(extensionValue.value, type, extensionValue.name, extensionValue.merge);
-    } else if (Array.isArray(value)) {
-      this.addReferences(value, type);
-    } else {
-      const useId = value.id || id;
+      const extensionValue = this._findExtensionValue(value);
+      value = extensionValue.value;
+    }
+
+    Object.entries(value).forEach(([id, customization]) => {
       const setName =
-        (type === CustomizationType.Global && 'setGlobalCustomization') ||
-        (type === CustomizationType.Default && 'setDefaultCustomization') ||
+        (type === CustomizationScope.Global && 'setGlobalCustomization') ||
+        (type === CustomizationScope.Default && 'setDefaultCustomization') ||
         'setModeCustomization';
-      this[setName](useId as string, value, merge);
-    }
+      this[setName](id as string, customization as Customization);
+    });
   }
 
   /**
    * Customizations can be specified as an array of strings or customizations,
    * or as an object whose key is the reference id, and the value is the string
    * or customization.
    */
-  addReferences(references?, type = CustomizationType.Global): void {
+  addReferences(references?: any, type = CustomizationScope.Global): void {
     if (!references) {
       return;
     }
     if (Array.isArray(references)) {
       references.forEach(item => {
-        this.addReference(item, type);
+        this._addReference(item, type);
       });
     } else {
-      for (const key of Object.keys(references)) {
-        const value = references[key];
-        this.addReference(value, type, key);
-      }
+      this._addReference(references, type);
     }
   }
 }
 
-/**
- * Custom merging function, to handle merging arrays and copying functions
- */
-function appendCustomizer(obj, src) {
-  if (Array.isArray(obj)) {
-    const srcArray = Array.isArray(src);
-    if (srcArray) {
-      return obj.concat(...src);
-    }
-    if (typeof src === 'object') {
-      const newList = obj.map(value => cloneDeepWith(value, cloneCustomizer));
-      for (const [key, value] of Object.entries(src)) {
-        const { position, isMerge } = findPosition(key, value, newList);
-        if (isMerge) {
-          if (typeof obj[position] === 'object') {
-            newList[position] = mergeWith(
-              Array.isArray(newList[position]) ? [] : {},
-              newList[position],
-              value,
-              appendCustomizer
-            );
-          } else {
-            newList[position] = value;
-          }
-        } else {
-          newList.splice(position, 0, value);
-        }
-      }
-      return newList;
-    }
-    return obj.concat(src);
+/** Add custom $filter command */
+extend('$filter', (query, original) => {
```

Comment:
Is there some way to use the customization service itself to extend the customization service $ matches?